src/win/tty.c

```diff
@@ -2255,25 +2250,67 @@ int uv_tty_reset_mode(void) {
   return 0;
 }
 
+static int uv__set_vtp(HANDLE handle, BOOL vtp) {
+  DWORD dwMode = 0;
+
+  if (!GetConsoleMode(handle, &dwMode)) {
+    return uv_translate_sys_error(GetLastError());
+  }
+
+  if (vtp) {
+    dwMode |= ENABLE_VIRTUAL_TERMINAL_PROCESSING;
+  } else {
+    dwMode &= ~(ENABLE_VIRTUAL_TERMINAL_PROCESSING);
+  }
+  if (!SetConsoleMode(handle, dwMode)) {
+    return uv_translate_sys_error(GetLastError());
+  }
+
+  return 0;
+}
+
 /* Determine whether or not this version of windows supports
  * proper ANSI color codes. Should be supported as of windows
  * 10 version 1511, build number 10.0.10586.
  */
 static void uv__determine_vterm_state(HANDLE handle) {
-  DWORD dwMode = 0;
-
-  if (!GetConsoleMode(handle, &dwMode)) {
-    uv__vterm_state = UV_UNSUPPORTED;
+  if (uv__set_vtp(handle, TRUE)) {
+    uv__vterm_state = UV_TTY_LEGACY;
     return;
   }
 
-  dwMode |= ENABLE_VIRTUAL_TERMINAL_PROCESSING;
-  if (!SetConsoleMode(handle, dwMode)) {
-    uv__vterm_state = UV_UNSUPPORTED;
-    return;
+  uv__vterm_state = UV_TTY_VTP;
+}
+
+int uv_tty_set_vterm_state(uv_tty_t* tty, uv_tty_vtermstate_t state) {
+  int ret = 0;
+
+  if (!(tty->flags & UV_HANDLE_WRITABLE)) {
+    return UV_EINVAL;
   }
 
-  uv__vterm_state = UV_SUPPORTED;
+  uv_sem_wait(&uv_tty_output_lock);
+  switch (state) {
+    case UV_TTY_AUTODETECT:
+      uv__determine_vterm_state(tty->handle);
+      break;
+    case UV_TTY_VTP:
+      if (!(ret = uv__set_vtp(tty->handle, TRUE))) {
+        uv__vterm_state = UV_TTY_VTP;
+      }
+      break;
+    case UV_TTY_LEGACY:
+      if (!(ret = uv__set_vtp(tty->handle, FALSE))) {
+        uv__vterm_state = UV_TTY_LEGACY;
+      }
```

Comment:
I think we should set `uv__vterm_state = UV_TTY_LEGACY` unconditionally, even if the user gave us a HANDLE which doesn't have permission to query the console mode. (I think the same is true of UV_TTY_VTP)

Reply:
The usage of `uv_tty_set_vterm_state` assumed by me is as follows.

```C
const char* term;
if (!strcmp(getenv("ConEmuAnsi"), "ON")) {
  term = "ConEmu";
  uv_tty_vterm_state(tty, UV_TTY_ANSI);
} else {
  int ret = uv_tty_vterm_state(tty, UV_TTY_VTP);
  assert(!ret || ret == UV_EINVAL);
  if (ret) {
    term = "Legacy";
  } else {
    term = "VTP";
  }
}
```

With this usage it is not necessary to call `SetConsoleMode` on the application side. Therefore I think that the validity of HANDLE should be confirmed. In that case, we should return `UV_EBADF` if the HANDLE verification fails. Or we should delete `uv_tty_t*` from the argument of `uv_tty_set_vterm_state` and use `uv__tty_console_handle`.

In the case of the legacy console, since the control sequence is outputted, even if VTP fails to turn on, I can not agree to set `uv__vterm_state = UV_TTY_VTP`. 

Since VTP is properly set when calling `uv_tty_init`, I can not find any other benefits. Besides that, what is the advantage of setting `ENABLE_VIRTUAL_TERMINAL_PROCESSING` with `uv_tty_set_vterm_state` you think?

Reply:
In this case, you don't actually know if VTP is on or not. You could have a valid HANDLE, but which doesn't allow you to inspect and/or change the setting. I would rather not fail in that case if we don't need to.

That seems like the job of `UV_TTY_AUTODETECT`, telling it it should use `VTP` when you meant `AUTODETECT` seems broken. If you want a getter, just add that method too.

For some background on my motivation, one reason I've needed to disable all of this logic right now for JuliaLang is that it doesn't gracefully handle the fact that sometimes some other programs may modify or control this state. So I'm pushing to make sure we don't make the same bad assumptions here.

Reply:
>  it doesn't gracefully handle the fact that sometimes some other programs may modify or control this state.

I do not think that JuliaLang has any meaning to disable this feature of libuv for the following reasons. Each console handle has VTP status. Also, `uv__vterm_state` is isolated for each process. So I think the problem is only when the program on JuliaLang calls `SetConsoleMode`. And in that case, the program should know what it is doing. Also, JuliaLang can hook `SetConsoleMode`.

Therefore, I think that this pull request does not make sense to change the VTP status of the console.

Reply:
>   when the program on JuliaLang calls SetConsoleMode. And in that case, the program should know what it is doing. Also, JuliaLang can hook SetConsoleMode. 

can you explain what you mean here in more detail?